doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,396 @@
+| `ALLOC_CONSOLE_MODE_NO_WINDOW`  | Allocate a console session _without_ a window, even if this process was created with `CREATE_NEW_WINDOW` or `DETACHED_PROCESS` |
+
+###### Notes
+
+Applications seeking backwards compatibility are encouraged to delay-load `AllocConsoleWithOptions` or check for its presence in
+the `api-ms-win-core-console-l1` APISet.
+
+## Inspiration
+
+Fusion manifest entries are used to make application-scoped decisions like this all the time, like `longPathAware` and
+`heapType`.
+
+CUI applications that can spawn a UI (or GUI applications that can print to a console) are commonplace on other
+platforms because there is no subsystem differentiation.
+
+## UI/UX Design
+
+There is no UI for this feature.
+
+## Capabilities
+
+### Accessibility
+
+This should have no impact on accessibility.
+
+### Security
+
+One reviewer brought up the potential for a malicious actor to spawn an endless stream of headless daemon processes.
+
+This proposal in no way changes the facilities available to malicious people for causing harm: they could have simply
+used `IMAGE_SUBSYSTEM_WINDOWS_GUI` and not presented a UI--an option that has been available to them for 35 years.
+
+### Reliability
+
+This should have no impact on reliability.
+
+### Compatibility
+
+An existing application opting into **detached** may constitute a breaking change, but the scope of the breakage is
+restricted to that application and is expected to be managed by the application.
+
+All behavioral changes are opt-in.
+
+> **EXAMPLE**: If Python updates python.exe to specify an allocation policy of **detached**, graphical python applications
+> will become double-click runnable from the graphical shell without spawning a console window. _However_, console-based
+> python applications will no longer spawn a console window when double-clicked from the graphical shell.
+>
+> In addition, if python.exe specifies **detached**, Console APIs will fail until a console is allocated.
+
+Python could work around this by calling [`AllocConsole`] or [new API `AllocConsoleWithOptions`](#allocconsolewithoptions)
+if it can be detected that console I/O is required.
+
+#### Downlevel
+
+On downlevel versions of Windows that do not understand (or expect) this manifest field, applications will allocate
+consoles as specified by their image subsystem (described in the [abstract](#abstract) above).
+
+### Performance, Power, and Efficiency
+
+This should have no impact on performance, power or efficiency.
+
+## Potential Issues
+
+### Shell Hang
+
+I am **not** proposing a change in how shells determine whether to wait for an application before returning to a prompt.
+This means that a console subsystem application that intends to primarily present a UI but occasionally print text to a
+console (therefore choosing the **detached** allocation policy) will cause the shell to "hang" and wait for it to
+exit.

```

Comment:
I think a shell's default wait for a console app should wait on the child process handle for a second and then, if the child is still running, check `GetConsoleProcessList()`. If the child process ID is in the list, then continue the wait loop. Alternatively, the console API could provide a `WaitForConsoleDetach()` function that simplifies this into a single call. It would return as soon as the process either terminates or calls `FreeConsole()`. Explicit waiting until child process termination would still be possible via something like CMD's `start /wait` command.